include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeSubsetSelection2D.h

```diff
@@ -0,0 +1,298 @@
+		// give the algorithm for this type. They just give the transformation but do not say
+		// what the transformation does so it is not clear until you implement it.
+		//
+		// This is a super confusing part of the paper, please kids, do not be like Bringmann et al.
+		// Keep it simple, stupid. Sometimes an additional index does not hurt.
+		//
+		//the algorithm works by inserting functions f_1 to f_i one-by-one, figuring out which functions
+		// are dominated (not being part of the upper envelope) and removing all functions which for
+		// function values x_i,x_i+1,... are already smaller than one of the other function.
+		// using the ordering relations given the set s contains the function ordered by (current) function value.
+		// so after iteration i we can just extract the largest function value for x_i by looking at the first element of s.
+		for (std::size_t i = 0; i != n; ++i) {
+
+			// remove dominated functions.
+			// as we push back into s,
+			// at the end of s are the functions with largest slope.
+			// therefore if we have the last two elements as s_-1 and s_-2 and the new
+			// function f, knowing that the intersection of s_-1 and f is smaller than the intersection
+			// of s_-1 and s_-2 means that s_-1 is dominated  by s_-2 and f and thus can be removed.
+			while (s.size() > 1 ) {
+				
+				double d1 = Intersection(functions[i], s.end()[-1]);
+				double d2 = Intersection(s.end()[-2], s.end()[-1]);
+
+				if (d1 <= d2 || std::abs(d1-d2) < 1.e-10) {//check for numeric stability
+					s.pop_back();
+				} else {
+					break;
+				}
+			}
+			//include the new function and store its index.
+			s.push_back(functions[i]);
+			s.back().index = i;
+			// at the beginning of s are the functions with smallest slope
+			// if the first function in s has a smaller function value for the current 
+			// x_i than the second function,
+			// we can safely remove it as it can not be part of the envelope any more
+			// (We are only looking at function values >=x from now on and thus the larger slope domintes)
+			while (s.size() > 1) {
+				double d1 = s[0].eval(points[i].f1);
+				double d2 = s[1].eval(points[i].f1);
+
+				if (d1 < d2 || std::abs(d1-d2) < 1.e-10) {
+					s.pop_front();
+				} else {
+					break;
+				}
+			}
+			//assign maximum
+			//the functions in s are ordered by function value  
+			// the function with the largest function value is currently at the front
+			h[i] = s[0].eval(points[i].f1);
+			chosen[i] = s[0].index;
+		}
+		return std::make_pair(std::move(h),std::move(chosen));
+	}
+	
+	
+	/// Fast calculation O(n*k) for the hypervolume selection problem
+	/// returns the selected points
+	void hypSSP(std::vector<Point>& front,std::size_t k)const{
+		std::size_t n = front.size();
+		std::vector<LinearFunction> functions(n);
+		
+		std::vector<std::vector<std::size_t> > chosen;
+		std::vector<double>  h(n,0.0);
+		for(std::size_t j=0; j != k-1; ++j) {//compute until k-1 elements are chosen
```

Comment:
Replacing j != k-1 with j+1 != k is not subject to underflow (but it does not solve the problem if k==0)